pkg/api/v1alpha1/cloudstackmachineconfig_types.go

```diff
@@ -40,6 +45,61 @@ type CloudStackMachineConfigSpec struct {
 	UserCustomDetails map[string]string `json:"userCustomDetails,omitempty"`
 }
 
+type CloudStackResourceDiskOffering struct {
+	CloudStackResourceIdentifier `json:",inline"`
+	// path the filesystem will use to mount in VM
+	MountPath string `json:"mountPath"`
+	// device name of the disk offering in VM, shows up in lsblk command
+	Device string `json:"device"`
+	// filesystem used to mkfs in disk offering partition
+	Filesystem string `json:"filesystem"`
+	// disk label used to label disk partition
+	Label string `json:"label"`
+}
+
+func (r *CloudStackResourceDiskOffering) Equal(o *CloudStackResourceDiskOffering) bool {
+	if r == o {
+		return true
+	}
+	if r == nil || o == nil {
+		return false
+	}
+	if r.Id != o.Id {
+		return false
+	}
+
+	if r.MountPath != o.MountPath || r.Filesystem != o.Filesystem || r.Label != o.Label || r.Device != o.Device {
+		return false
+	}
+	return r.Id == "" && o.Id == "" && r.Name == o.Name
+}
+
+func (r *CloudStackResourceDiskOffering) Validate() (err error, field string, value string) {
+	if r.Provided() {
+		if len(r.MountPath) < 2 || !strings.HasPrefix(r.MountPath, "/") {
+			return errors.New("must be non-empty and starts with /"), "mountPath", r.MountPath
+		}
+		if len(r.Filesystem) < 1 {
+			return errors.New("empty filesystem"), "filesystem", r.Filesystem
+		}
+		if len(r.Device) < 1 {
+			return errors.New("empty device"), "device", r.Device
+		}
+		if len(r.Label) < 1 {
+			return errors.New("empty label"), "label", r.Label
+		}
+	} else {
+		if len(r.MountPath) + len(r.Filesystem) + len(r.Device) + len(r.Label) > 0 {
+			return errors.New("empty id/name"), "id or name", r.Id
+		}
+	}
+	return nil, "", ""
+}
+
+func (r *CloudStackResourceDiskOffering) Provided() bool {
```

Comment:
do we need this? :0

Reply:
trying to re-use this logic in validateMachineConfig in validator.go, so this simple method is created.

Reply:
I feel this method name is confusing. What does it mean by Provided? I need to go back and see `len(r.Id) > 0 || len(r.Name) > 0`. imo it's more clear just to use the if statement. But your call